source/common/event/dispatcher_impl.cc

```diff
@@ -121,15 +122,29 @@ DispatcherImpl::createClientConnection(Network::Address::InstanceConstSharedPtr
                                                          std::move(transport_socket), options);
 }
 
+absl::optional<bool> DispatcherImpl::use_apple_api_for_dns_lookups = absl::nullopt;
+
 Network::DnsResolverSharedPtr DispatcherImpl::createDnsResolver(
     const std::vector<Network::Address::InstanceConstSharedPtr>& resolvers,
     const bool use_tcp_for_dns_lookups) {
   ASSERT(isThreadSafe());
 #ifdef __APPLE__
-  if (Runtime::LoaderSingleton::getExisting()->snapshot().getBoolean(
-          "dns.use_apple_api_for_dns_lookups", false)) {
-    UNREFERENCED_PARAMETER(resolvers);
-    UNREFERENCED_PARAMETER(use_tcp_for_dns_lookups);
+  if (!DispatcherImpl::use_apple_api_for_dns_lookups.has_value()) {
+    use_apple_api_for_dns_lookups =
+        Runtime::runtimeFeatureEnabled("envoy.reloadable_features.use_apple_api_for_dns_lookups");
+  }
+
+  if (DispatcherImpl::use_apple_api_for_dns_lookups.value()) {
+    if (!resolvers.empty()) {
+      ENVOY_LOG(
+          warn,
+          "defining custom resolvers is not possible when using Apple APIs for DNS resolution");
+    }
+    if (use_tcp_for_dns_lookups) {
+      ENVOY_LOG(
+          warn,
+          "using TCP for DNS lookups is not possible when using Apple APIs for DNS resolution");
+    }
```

Comment:
Please add tests for these. I could go either way on warn vs. outright failure. Outright failure seems potentially better tbh.

Reply:
Went for failure, that way I also get to test with my favorite expectation (`EXPECT_DEATH`)

Reply:
The only, slight, concern I have is about potentially failing mac setups that are running master, and _also_ have used these config options... although as you have pointed out it is probably unlikely?